### ISSUE TITLE
JsonSerializer deserialize fails when attempting to deserialize object[] arrays

### ISSUE THREAD

**Issue by pranavkm** [dotnet/corefx#36167]
Consider:

```C#
var content = "[1]";
var deserialized = JsonSerializer.Parse<object[]>(content);
Console.WriteLine(deserialized);
```

This fails with:

```C#
Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at System.Text.Json.Serialization.JsonPropertyInfo`2.Read(JsonTokenType tokenType, JsonSerializerOptions options, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.JsonSerializer.HandleValue(JsonTokenType tokenType, JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.Serialization.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.Serialization.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonSerializer.Parse[TValue](String json, JsonSerializerOptions options)
```

Tested using shared runtime **3.0.0-preview4-27519-10**

**Comment by steveharter** [comment]
This is current by design until we support deserializing `object` into `JsonElement`. In the example above, we don't know whether "1" is a short, int, long, decimal, etc so we plan on deserializing into a `JsonElement` where the token type and convenience methods will be provided to obtain the expected type.

**Comment by steveharter** [comment]
Related to https://github.com/dotnet/corefx/issues/36169

**Comment by BrennanConroy** [comment]
We have use cases where we will be given `object` as the type to be deserialized. We can't return a `JsonElement` back to the user in this case.

**Comment by steveharter** [comment]
> We have use cases where we will be given object as the type to be deserialized. We can't return a JsonElement back to the user in this case.

What is the expected return type?

**Comment by BrennanConroy** [comment]
>What is the expected return type?

An object. Basically what this method does https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonSerializer_Deserialize.htm

**Comment by JeremyKuhne** [comment]
Validated that this was fixed and added a specific test in #38029.
